[5.1] Support opaque stored property types.

jckarter opened the pull request:
Explanation: Allows stored properties with initializers to have opaque types, with the underlying type inferred from the initializer. Adds an error when a stored property has an opaque type but no initializer (where we used to silently accept this and crash later).

Scope: Missing part of opaque types feature

Issue: rdar://problem/51127135

Risk: Low

Reviewed by: @slavapestov, @xedin

jckarter committed:
Support opaque stored property types.

Type-check the initializer expression to get the underlying type (or raise an error if there
is none).

jckarter commented:
@swift-ci Please test

slavapestov commented on include/swift/AST/DiagnosticsSema.def:
```diff
@@ -3310,6 +3310,9 @@ ERROR(trailing_closure_requires_parens,none,
       " context", ())
 
+ERROR(opaque_type_var_no_init,none,
+      "property declares an opaque return type, but has no initializer "
```
Do we want to call these 'initial value' or 'initializer expression' in diagnostics?

slavapestov commented on test/type/opaque.swift:
```diff
@@ -418,3 +416,14 @@ func testCoercionDiagnostics() {
   opaqueOpt = () // expected-error {{cannot assign value of type '()' to type '(some P)?'}} {{none}}
 }
+
+var globalVar: some P = 17
```
Can local variables have opaque result types?

jckarter replied:
Yeah.

jckarter committed:
Followup fixes for stored opaque-type properties.

- TypedPatterns expect to have a contextual type in generic contexts.
- mapTypeOutOfContext failed for NestedArchetypes under opaque archetypes.

jckarter commented:
@swift-ci Please test
